### ISSUE TITLE
Tab character incorrectly assumed to be composed of 8 characters

### ISSUE THREAD

**Issue by PiJoKra** [haskell/haskell-language-server#1987]
### Your environment
Output of `haskell-language-server --probe-tools` or `haskell-language-server-wrapper --probe-tools`:

```sh
haskell-language-server version: 1.0.0.0 (GHC: 8.10.4) (PATH: [MYHOME]/.ghcup/bin/haskell-language-server-wrapper-1.0.0) (GIT hash: 4cd1cf934638881e52b3eba9f70157a4b799c0e9)
Tool versions found on the $PATH
cabal:          3.4.0.0
stack:          2.7.1
ghc:            8.6.5

```
Which OS do you use: Ubuntu

Which lsp-client do you use: Vim with Coc

Describe your project (alternative: link to the project): stack-project

Contents of `hie.yaml`:
```yaml
cradle:
  stack:
      component: "name-of-project:lib"
```

### Steps to reproduce

Type the following code
```haskell
module Test where

main :: IO ()
main = do
	putStr "hello" -- there is a tab-character (\t) in front of putStr
```
And request information about `putStr`
### Expected behaviour

I should get a pop-up with type information and a short description.

### Actual behaviour

I get a pop-up about `IO` instead. 
If I put my cursor further ahead, over the string "hello", I do get information about `putStr`. It seems HLS treats a tab as 8 separate characters. Setting a tab-size of 8 does not help, since it is still one character.
### Include debug information
Paste the logs from the lsp-client, e.g. for [VS Code](https://github.com/haskell/vscode-haskell#troubleshooting)

<details>
<summary>
LSP logs:
</summary>
The first part is when the cursor is on `putStr` and gives information about `IO`, the second part is when the cursor is on the last part of "hello" and gives information about `putStr`

File: [hls-output.txt](https://github.com/haskell/haskell-language-server/files/6727553/hls-output.txt)

Gist: [hls-output.txt gist](https://gist.github.com/PiJoKra/42d94188632e3343a83f9e98a24bd149)

Please let me know if you need more of the output.
</details>

**Comment by joeyh** [comment]
I am also hit by this bug. Using emacs here. (I've disabled the usual emacs indent stuff since I do use tabs.)

**Comment by michaelpj** [comment]
I think this may be ultimately related to https://github.com/haskell/haskell-language-server/issues/2646 and our broken handling of unicode.

**Comment by ag-eitilt** [comment]
I'm inclined to agree with the last few people above in believing this is the fault of GHC/tooling between it and HLS -- directly running Cabal on the code also reports column numbers based on that 8-column width.  The next question is then whether the patch recalculating that should be written here (since 8 columns is the canonical depth for calculating indentation-sensitive blocks, GHC reporting the wide value makes sense) or deeper in the stack (since it's ultimately reporting a misleading value for most contexts outside of depth calculation, given modern editor-config trends).

For the record, I think it's *probably* better for HLS to be the one to recalculate it, particularly given there already needs to be some massaging for the mentioned #2646.  That specific issue is ironically almost the opposite of this one (this is one character being counted as many when it should only be one; that is one character *needing* to be counted as many when it *is* reported as one), but it is ultimately stemming from the same root cause of Haskell character counting being different from the LSP expectations.  I just think it's probably a good idea to have it on record, from people who understand the full ecosystem better than I do, that it is a fix to be applied within the HLS layer before anyone goes digging into the code anywhere.[^1]

[^1]: I'm not able to commit to being the one to do so quite yet; work's busy for the near future, and I'm still in the middle of restoring my personal systems after a sweeping OS reinstall.  If nobody jumps on these in the next couple months, I'll probably see what I can do eventually since the LSP mishighlighting is indeed annoying, but if anyone has more initiative than I do, don't let me stop you.

**Comment by ag-eitilt** [comment]
Yeah, looking through the code, https://github.com/haskell/haskell-language-server/issues/2646#issuecomment-1024990401 is absolutely correct.  There's a /lot/ of `SrcSpan` (code points) throughout the codebase, which are blindly converted to `Range` (LSP-expected UTF-16), and the *only* place which performs actual UTF-16 conversion (the [semantic tokens plugin](https://github.com/haskell/haskell-language-server/blob/9e6c6c1fbec8333a05196c693c7e9250359990be/plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs#L129)) is the only one that properly respects tabs.  There might be some additional work necessary depending on if we need to offset [parsed column numbers](https://github.com/haskell/haskell-language-server/blob/9e6c6c1fbec8333a05196c693c7e9250359990be/ghcide/src/Development/IDE/Types/Location.hs#L81),[^1] but otherwise the bulk of fixing this seems like it will be served by fixing that more broadly.

[^1]: The semantic tokens implementation doesn't do anything special about tabs, but nevertheless all highlighting is in the right place.  Makes me think that the `SrcSpan` data *isn't* normalized to `\t` == 8 cols, and that that's just something "helpful" GHC is doing when printing the error message.  If so, that's quite annoying since it means there are two categories of `SrcSpan` with subtly different semantics floating around.  Though since the only code which uses `readSrcSpan` is [the](https://github.com/haskell/haskell-language-server/blob/9e6c6c1fbec8333a05196c693c7e9250359990be/plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs#L402) [refactor](https://github.com/haskell/haskell-language-server/blob/9e6c6c1fbec8333a05196c693c7e9250359990be/plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs#L844) [plugin](https://github.com/haskell/haskell-language-server/blob/9e6c6c1fbec8333a05196c693c7e9250359990be/plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs#L987), maybe we're lucky and that's a dead code path which can be trimmed entirely.

The code diving has reinforced my belief that this is something that should be handled *somewhere* in this repo; probably in `ghcide`, though each of the plugins will need to be checked for proper behaviour.
